### PR TITLE
Expose literal userspace lifecycle hooks for projects

### PR DESCRIPTION
**Pull request by jonastemplestein**
## Summary

This PR makes the editable config-repo `worker.ts` the literal userspace place
for arbitrary project lifecycle code.

There is no lifecycle framework, `reconciliation-requested` event, desired-state
format, userspace creation handshake, or hidden “configuration processor”.
The existing worker entrypoint receives ordinary events and makes ordinary ITX
calls in its existing `processEvent` switch:

```ts
protected override async processEvent(event: StreamEvent): Promise<void> {
  switch (event.type) {
    case "events.iterate.com/project/heartbeat-triggered": {
      if (event.path !== "/") break;
      console.log("Project heartbeat fired", {
        scheduleKey: event.payload?.scheduleKey,
      });
      // Put arbitrary periodic ITX code here.
      break;
    }

    case "events.iterate.com/stream/woken": {
      if (event.path !== "/") break;
      // Put arbitrary root-stream wake / OS-deploy code here.
      break;
    }

    case "events.iterate.com/project/worker-updated": {
      if (event.path !== "/") break;
      // The platform has built, loaded, and probed the current config worker.
      // Put arbitrary idempotent ITX code directly here.
      const itx = await this.itx;
      await itx.scheduler.set({
        key: "iterate/config/heartbeat/every-15-minutes",
        recurrence: { every: 15 * 60 },
        script: `async (itx, schedule, trigger) => {
          await itx.streams.get("/").append({
            type: "events.iterate.com/project/heartbeat-triggered",
            idempotencyKey: "iterate/config/heartbeat:" + trigger.executionId,
            payload: { scheduleKey: schedule.key },
          });
        }`,
      });
      break;
    }

    default:
      break;
  }

  await this.#aiLintApp.processEvent(event);
  await this.#guestbookApp.processEvent(event);
}
```

To install another schedule, subscription, capability, or event, write the ITX
call directly in one or more of those cases. Shared behavior can be an ordinary
local function. Nothing parses `worker.ts` as configuration.

## Lifecycle events

| Event | Durable meaning | Intended use |
| --- | --- | --- |
| Root `project/worker-updated` | The platform built, loaded, invoked, and successfully probed the current default worker. Payload: `{ commitOid }`, where the commit is the source the platform actually served. | Apply the repo’s current executable configuration with direct, idempotent ITX calls. |
| Root `project/worker-update-failed` | A post-creation config commit deterministically failed to build. Payload: `{ commitOid, error }`. | Durable diagnosis; a later fixed commit proceeds independently. |
| Root `stream/woken` | A new incarnation of the root Stream Durable Object started. | Run code when the project stream wakes after hibernation or an OS deployment. |
| Root `project/heartbeat-triggered` | One named project-owned Scheduler execution appended its userspace event. Payload: `{ scheduleKey }`. | Periodic idempotent checking or repair. |

`project/create-requested` and `project/created` are deliberately absent from
the userspace switch. They are facts in the platform creation saga. Code added
to the config repo next week cannot usefully depend on a creation event that
happened once last week.

### Config events are copied to the project root

The config repo owns one normal source subscription, using the subscription
model now on `main`:

```ts
{
  type: "events.iterate.com/stream/subscription-configured",
  payload: {
    subscriptionKey: "project-config-to-root",
    receiver: {
      action: "copy-to-stream",
      receivingStreamPath: "/",
      delivery: {
        start: "now",
        onFailingEvent: "halt",
      },
    },
  },
}
```

The source stream owns the cursor. Copied events arrive on `/` with
`source.copiedFrom` provenance. The Project processor accepts only the exact
`project-config-to-root` copy from this project’s `/repos/config` stream; this
repo-specific topology stays in platform code and never leaks into the
userspace switch.

### Project creation remains one conventional saga

```text
project/create-requested
  -> create and reduce the root capability host, primary scheduler,
     config repo, and email router
  -> receive the copied config-repo repos/created terminal
  -> wait for the trusted seeded config worker to build, load, and answer
  -> atomically install the permanent root project-worker feed
     and append project/created + the first project/worker-updated

project/create-requested -> project/created | project/create-failed
```

The root worker feed is intentionally absent while the seed worker builds, so
the Stream Durable Object cannot diagnose a not-yet-built worker as a broken
subscriber. Once the worker answers, the saga installs this literal receiver:

```ts
{
  type: "events.iterate.com/stream/subscription-configured",
  payload: {
    subscriptionKey: "project-worker",
    receiver: {
      action: "itx-call",
      expression: ["processEventBatch"],
      delivery: {
        start: "now",
        onFailingEvent: "skip",
      },
    },
  },
}
```

`start: "now"` anchors the cursor at the subscription configuration event. It
skips the preceding platform bootstrap history, including
`project/create-requested`; the following terminal and lifecycle facts remain
ordinary feed events. The template simply has no creation-event cases.

The feed configuration, `project/created`, and first `project/worker-updated`
are one append batch. Therefore `project/created` has a precise meaning:
sibling processors exist, the seeded default worker answered, and its durable
root feed was committed. Creation does not wait for userspace to consume the
worker-update event.

The two creation terminals use canonical
`iterate-internal/project-creation-terminal` idempotency keys and the Project
Durable Object commits them through the server-only Stream stub, fenced to the
stream lifetime that delivered the saga event. Public appends reject those keys
before idempotency lookup, so a caller cannot forge a terminal or reserve its
canonical key. The reducer also matches the exact open request. This boundary
is confined to the platform creation saga; userspace receives ordinary events
and no authority API.

Deterministic config-repo or seed source-build failures append
`project/create-failed`. Transient repo/build availability and an in-progress
build leave the Project processor cursor open for durable redelivery. The old
`project/ready`, `waitUntilReady`, and `ready` deployment state are removed;
the public API waits for `created` or `create-failed`.

### Later config commits produce a worker lifecycle fact

```text
/repos/config repo/commit-completed
  -> project-config-to-root copies the event onto /
  -> ProjectProcessor validates source.copiedFrom
  -> probe the authoritative current worker through DynamicWorkerRunner.fetch
       still building / transiently unavailable -> keep cursor open; redeliver
       deterministic source failure             -> project/worker-update-failed
       worker answers                            -> project/worker-updated
  -> the current userspace worker receives the lifecycle event
  -> its literal switch runs direct ITX code
```

The platform fetch lane, not `itx.worker.fetch`, performs the probe. It resolves,
builds, loads, and invokes the repo-backed worker, then removes any
userspace-authored serve header and stamps the commit it actually served.

This is intentionally a reconcile-current-config hook rather than an exact
per-artifact deployment callback. If commits A and B land quickly, B may already
be HEAD while the Project processor handles A’s copied repo event. The resulting
`project/worker-updated` stays idempotently keyed to trigger A but carries B—the
worker the platform actually proved runnable. Replaying A therefore finds its
committed outcome instead of probing a later HEAD and contradicting that result.
B’s own trigger may run the idempotent userspace hook again; that is deliberate,
because lifecycle code is required to be idempotent.

For deterministic post-creation failure, the failure fact is keyed to the
triggering commit. A later config commit can repair the worker without reopening
project creation. Availability failures do not become false domain failures.

## Heartbeats and Scheduler behavior

The seeded worker installs one schedule using the Scheduler’s existing keyed
shape:

```ts
await itx.scheduler.set({
  key: "iterate/config/heartbeat/every-15-minutes",
  recurrence: { every: 900 },
  script: `async (itx, schedule, trigger) => {
    await itx.streams.get("/").append({
      type: "events.iterate.com/project/heartbeat-triggered",
      idempotencyKey: "iterate/config/heartbeat:" + trigger.executionId,
      payload: { scheduleKey: schedule.key },
    });
  }`,
});
```

The one payload property distinguishes this schedule from future heartbeats:

```ts
{ scheduleKey: "iterate/config/heartbeat/every-15-minutes" }
```

Multiple schedules are multiple literal `set` calls with different keys. A
fast test project can use `{ every: 1 }`; a project that wants no heartbeat
deletes the call before creating/applying that config. Removing an already
applied call does not silently delete durable state—cleanup is the explicit
`scheduler.cancel(key)` call.

This PR makes `scheduler.set(...)` itself idempotent for an unchanged canonical
definition. An exact match returns the existing schedule without appending
another `scheduler/schedule-set` event or resetting `nextTriggerAt`, `runCount`,
`definedAtOffset`, or `setAt`. A changed or absent definition retains the normal
keyed replacement behavior. Concurrent writes are serialized so simultaneous
identical calls cannot both pass the read-before-append check. There is no
separate `scheduler.ensure(...)`.

Interval schedules coalesce downtime into the next occurrence. They do not
backfill one event for every missed 15-minute slot. Each real execution uses its
execution ID as the heartbeat append’s idempotency key, so an at-least-once
script retry cannot append the same heartbeat twice.

## `this.itx`

`IterateWorkerEntrypoint` now exposes protected `await this.itx`. It memoizes
`env.ITX.get()` on the stateless entrypoint instance, so a fetch or event batch
can reuse one project-root session without repeating the acquisition ceremony.
Cloudflare releases the invocation’s RPC stubs when that execution context
ends.

The Durable Object base class deliberately does not expose the same memoization:
Durable Objects span many invocations and Cloudflare provides no eviction
callback on which to dispose a long-lived RPC session. Those call sites keep
explicit invocation-scoped acquisition.

## Problems solved

| Problem | Solution |
| --- | --- |
| Config code added after birth cannot depend on a one-time creation event. | Use recurring worker-update, root-wake, and heartbeat events. |
| Userspace would need a special `/repos/config` provenance `if`. | Copy config facts to `/` and translate the important one once in the platform Project processor. |
| `repo/commit-completed` does not prove the new worker built or can run. | Emit `project/worker-updated` only after the current worker answers a trusted platform probe. |
| Rapid commits could make a lifecycle fact claim the wrong runnable source or replay against a different HEAD. | Publish the commit actually served, but key the outcome to the triggering commit so replay never re-probes that trigger. |
| Lifecycle work was accumulating abstractions around a simple need. | Keep explicit lifecycle facts and direct ITX calls in the existing `worker.ts` switch. |
| The root worker feed could race the seed build and look broken. | Install it only after the seed worker answers, atomically with the creation terminal. |
| A public append could forge `source.processor` or reserve a creation-terminal key. | Use the existing internal-key boundary in the Stream platform, a server-only append, an exact-request check, and a stream-lifetime fence. |
| Re-running lifecycle code reset a matching schedule’s clock. | Make `scheduler.set` a no-op for an unchanged canonical definition. |
| Concurrent identical `set` calls could append duplicate definitions. | Serialize Scheduler mutations around compare-and-append. |
| Every config-worker hook repeated disposable `env.ITX.get()` ceremony. | Memoize `this.itx` for each stateless WorkerEntrypoint invocation. |
| Downtime could imply a burst of historical heartbeats. | Coalesce missed intervals; no backfill. |
| `project/ready` duplicated the creation terminal. | Use `create-requested -> created | create-failed`. |

## Non-goals and rollout

- No `reconciliation-requested` event or reconciliation framework.
- No userspace handling of `project/create-requested` or `project/created`.
- No project-creation API schema for heartbeat arrays.
- No implicit source scanning, desired-state diff, or implicit deletion.
- No separate userspace stateful processor.
- No userspace authority field or lifecycle-authentication API. This change uses the existing Stream platform internal-key boundary only for platform creation terminals.
- No backfill or compatibility migration for existing projects; they can be
  deleted and recreated for this rollout.

## Verification

- `pnpm install`
- `pnpm typecheck` (including the config-repo template)
- `pnpm lint`
- `pnpm format`
- `pnpm test`
  - OS: 240 files passed; 2,445 tests passed, 12 expected failures, 1 skipped
  - Project lifecycle: creation terminal, copy provenance, seed probe,
    current-served worker identity, deterministic failure, transient
    redelivery, lost acknowledgements, full replay
  - Scheduler: canonical equality, concurrent identical `set`, changed
    definitions, runtime-state preservation, trigger and cancellation
  - Streams: source-owned copy delivery and unavailable-receiver cursor
    behavior
  - Config template: literal switch, ignored creation facts, heartbeat payload,
    root wake, scheduler call, and memoized `this.itx`

## Environment Config Lease

<details>
<summary>No preview slot recorded.</summary>

| app | status | commit | preview | size (gzip) | deploy duration | test duration | retries | cleanup duration | workflow run | updated | summary |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Auth | released | `ccbfd6d` | [https://auth.iterate-preview-15.com](https://auth.iterate-preview-15.com) | 814.3 KiB | 35.7s | 13.4s |  | 4.1s | [Workflow run](https://github.com/iterate/iterate/actions/runs/1364602076573) | 2026-07-29T12:09:57.539Z | Preview app released. |
| Dummy Petshop | released | `ccbfd6d` | [https://dummy-petshop.iterate-preview-15.com](https://dummy-petshop.iterate-preview-15.com) | 198.3 KiB | 256ms | 6.7s |  | 1.1s | [Workflow run](https://github.com/iterate/iterate/actions/runs/1364602076573) | 2026-07-29T12:09:54.602Z | Preview app released. |
| OS | released | `ccbfd6d` | [https://os.iterate-preview-15.com](https://os.iterate-preview-15.com) | 4.92 MiB (-3.7 KiB vs main) | 74.7s | 203.2s |  | 20.4s | [Workflow run](https://github.com/iterate/iterate/actions/runs/1364602076573) | 2026-07-29T12:10:13.906Z | Preview app released. |
| Semaphore | released | `ccbfd6d` | [https://semaphore.iterate-preview-15.com](https://semaphore.iterate-preview-15.com) | 441.1 KiB | 28.1s | 64.7s |  | 708ms | [Workflow run](https://github.com/iterate/iterate/actions/runs/1364602076573) | 2026-07-29T12:09:54.175Z | Preview app released. |
| Streams Example App | released | `ccbfd6d` | [https://streams.iterate-preview-15.com](https://streams.iterate-preview-15.com) | 1.45 MiB | 29.4s | 87.4s |  | 11.1s | [Workflow run](https://github.com/iterate/iterate/actions/runs/1364602076573) | 2026-07-29T12:10:04.513Z | Preview app released. |

</details>

| Group | Lines | Significant |
| --- | ---: | ---: |
| Product | +1,347 -333 | +1,005 -185 🟩🟩🟩🟩🟥 |
| Mobile | +2 -2 | +2 -2 🟩⬜⬜⬜⬜ |
| UI components | +38 -10 | +32 -7 🟩⬜⬜⬜⬜ |
| Tests | +1,169 -225 | +1,048 -179 🟩🟩🟩🟩🟥 |
| CI & scripts | +1 -1 | +1 -1 🟩⬜⬜⬜⬜ |
| Docs | +215 -56 | +198 -56 🟩⬜⬜⬜⬜ |
| Other | +5 -6 | +0 -0 ⬜⬜⬜⬜⬜ |
| Generated | +612 -419 | +242 -117 🟩⬜⬜⬜⬜ |
| **Total** | **+3,389 -1,052** | **+2,528 -547** 🟩🟩🟩🟩🟥 |

<sub>Lines counts every changed line; Significant ignores blank lines, JS comments, and TypeScript lines with no runtime output. Between af77e56 and ccbfd6d, bucketed first-match-wins into the groups defined in `scripts/ci/loc-report.ts`.</sub>

---

> [!NOTE]
> **High Risk**
> Large change to project creation, worker delivery timing on the root stream, and repo head authority for builds—core platform paths with extensive test coverage but high blast radius if edge cases slip through.
> 
> **Overview**
> This PR reshapes **project bootstrap and config lifecycle** so userspace code lives in the config-repo `worker.ts` switch—no reconciliation framework, no handling of platform-only `project/create-requested` / `project/created`.
> 
> **Creation** now starts with `project/create-requested`, waits for the config repo and a **trusted** seed-worker readiness probe via `DynamicWorkerRunner`, then atomically commits the permanent root `project-worker` subscription, terminal `project/created`, and the first `project/worker-updated`. Deterministic repo or build failures land on `project/create-failed`; transient build/availability errors stay open for redelivery. Public APIs move from `waitUntilReady` / `ready` to **`waitUntilCreated`** and deployment states like `created`, `creating`, and `failed`.
> 
> **After creation**, copied `/repos/config` `repo/commit-completed` facts trigger another probe; success becomes `project/worker-updated` (outcome keyed to the triggering commit but carrying the commit actually served), deterministic failures become `project/worker-update-failed`. The seeded template reacts in `project/worker-updated` with a literal `scheduler.set` for a 15-minute heartbeat that appends `project/heartbeat-triggered`; root `stream/woken` is available for wake/deploy work.
> 
> **Scheduler** `set()` is a no-op when the canonical definition already matches (clock and run count preserved), with serialized writes to avoid duplicate `schedule-set` events. **Repo `getHead`** for worker source resolution rejects stale heads until Artifacts replicas converge, so lifecycle work cannot acknowledge through wrong code.
> 
> Dashboard creation progress and sidebar filtering follow the new terminal states and failure UI.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ccbfd6d6fdcfc3cc646f70de58d18e5d5c15f842. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: apps/os/src/domains/projects/project-processor-implementation.ts [hunk 1]
```diff
@@ -81,49 +92,162 @@ export class ProjectProcessor extends StreamProcessor<
...
+                  deliver: { afterOffset: createRequestedAtOffset },
+                  onPoison: "skip",
+                },
+              },
+              {
+                type: "events.iterate.com/project/created",
+                idempotencyKey: projectCreatedIdempotencyKey,
+                payload: { ...createRequest, createRequestedAtOffset },
+              },
+            ),
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/projects/project-processor-implementation.ts]: Fixed in e1f654566. The stream RPC trust boundary now rejects userspace `platform:` idempotency keys and every root `project-worker` subscription mutation (configure/remove/park/resume/seek). The creation saga uses that reserved namespace for its temporary configuration, removal, permanent configuration, and terminal facts, so newly widened project access cannot squat the keys or disturb the delivery fence. Added regression coverage proving the userspace calls never reach the Stream DO and platform-processor-stamped calls do; full tests/typecheck/lint/format/Knip are green.

**Comment by mmkal** [comment]
🤖 Heads-up from the session that landed #2307: this branch is 12 behind main and holds an active preview lease, which means its preview runs the **old** package pinning — freshly-created project workers there will fail to build (`Failed to resolve 'iterate/github-ai-linter' from worker.ts`), hanging every project-creation-dependent spec (same failure mode diagnosed on #2309, fixed by #2307 which is now on main).

Please merge main into this branch to pick up the fix. I attempted the merge but stopped at 8 conflicted files that are semantic to this branch's feature work (`rpc-targets.ts`, `project-root-redirect.*`, `specs/test-support/forged-session.ts`, config-repo-template files) — better resolved by this branch's owner. Note `config-repo-template.generated.ts` should be **regenerated**, not hand-merged. Until then, preview e2e failures about project readiness on this PR are not caused by this branch.

File: apps/os/src/domains/projects/project-processor-contract.ts [hunk 2]
```diff
@@ -419,12 +459,108 @@ export type ProjectProcessorContract = typeof ProjectProcessorContract;
...
+  if (
+    event.path !== "/" ||
+    event.source?.crossPostedFrom !== undefined ||
+    processor?.authority !== "platform" ||
+    processor.slug !== ProjectProcessorContract.slug ||
+    processor.version !== ProjectProcessorContract.version ||
+    processor.stream.path !== "/" ||
+    processor.stream.projectId !== projectId
+  ) {
+    return null;
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/projects/project-processor-contract.ts:569]: Fixed in a53a93216. `parseProjectCreationTerminal` no longer compares the stamped source version to the live contract version; platform authority, processor slug, root stream/project identity, while-processing provenance, idempotency key, payload, request, and offset remain validated. Added a regression proving a trusted `0.2.0` terminal still reduces under the current `0.3.0` contract. The full local gate is green.

File: apps/os/src/domains/projects/project-processor-implementation.ts [hunk 3]
```diff
@@ -607,6 +737,15 @@ export class ProjectProcessor extends StreamProcessor<
...
   itx: ProjectRpcTarget;
+  /** Internal root-stream cursor fence used only by the creation saga. */
+  waitUntilSubscriptionDelivered(input: {
+    configuredAtOffset: number;
+    eventType: string;
+    expression: ["processEventBatch"];
+    subscriptionKey: string;
+    targetOffset: number;
+    timeoutMs: number;
+  }): Promise<void>;
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/projects/project-processor-implementation.ts]: Fixed in a53a93216. `projectRecordFromState` now falls back from `birthCertificate.config.slug` to `createRequest.config.slug`, and only then to the raw project ID. The regression test appends the custom-domain request while `birthCertificate` is still null and verifies that `demo` from the creation request is passed to the provisioner. The full local gate is green.

File: apps/os/src/domains/repos/repo-durable-object.ts [hunk 4]
```diff
@@ -248,13 +250,17 @@ export class RepoDurableObject extends DurableObject<Env> {
...
+    // observed external push, must not be served even once. Otherwise the
+    // root feed could acknowledge repo/commit-completed through the previous
+    // config worker and permanently lose that revision's lifecycle reaction.
     const decision = decideHeadResolution(this.#branchAuthority(branch), head.commitOid);
-    if (!decision.cache) return { branch, ...head };
+    if (!decision.cache) {
+      throw new RepoNotSeededError(
+        `Repo branch "${branch}" is still converging after an observed push (resolved ${head.commitOid}; ${decision.reason}).`,
+      );
+    }
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/repos/repo-durable-object.ts:263]: Fixed in 938a68e0a. The sync/reset paths no longer call strict getHead() merely to warm the cache after a successful transfer. They record the adopted pushed head, invalidate the cache, and append repo/github-synced; the first real source consumer then waits for authoritative Artifacts convergence. This preserves the successful sync fact without allowing stale worker code to acknowledge downstream events.

File: apps/os/src/domains/projects/project-processor-implementation.ts [hunk 5]
```diff
@@ -81,50 +89,120 @@ export class ProjectProcessor extends StreamProcessor<
...
+              type: "events.iterate.com/project/create-failed",
+              idempotencyKey: "project/create-failed",
+              payload: {
+                createRequestedAtOffset,
+                error: `Default project worker bootstrap failed: ${errorMessage(error)}`,
+                request: createRequest,
+              },
+            });
+            return;
+          }
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/projects/project-processor-implementation.ts:242]: Fixed in 4c0c73a0e. Before either config-repo failure handling or the worker probe, the reaction now reads and validates both durable terminal idempotency keys. An existing project/create-failed returns without re-probing; an existing project/created preserves the prior lost-ack path; both terminals are an explicit invariant violation. The new fresh-cursor regression starts before the config certificate with the failed terminal already durable and proves zero re-probes, one failure, and no project/created.

File: apps/os/src/domains/projects/project-processor-implementation.ts [hunk 6]
```diff
@@ -81,53 +96,194 @@ export class ProjectProcessor extends StreamProcessor<
...
+          await append({
+            type: "events.iterate.com/project/worker-updated",
+            idempotencyKey: outcomeIdempotencyKey,
+            payload: { commitOid: event.payload.commitOid },
+          });
+        });
+        break;
+      }
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/projects/project-processor-implementation.ts:348]: Fixed in 424615b5d. The config-commit translation now exits while `state.birthCertificate` is null, so the trusted seed commit cannot probe or hold the ProjectProcessor cursor ahead of creation terminalization. The regression test drives `project/create-requested` → seed `repo/commit-completed` → config `repos/created` and proves there is exactly one worker probe (the creation probe), no `project/worker-updated`, and one `project/created`. The full local typecheck/lint/format/test gate passes.

File: apps/os/config-repo-template/worker.ts [hunk 7]
```diff
@@ -27,6 +27,43 @@ export default class ProjectWorker extends IterateWorkerEntrypoint {
...
+          recurrence: { every: 15 * 60 },
+          script: `async (itx, schedule, trigger) => {
+            await itx.streams.get("/").append({
+              type: "events.iterate.com/project/heartbeat-triggered",
+              idempotencyKey: "iterate/config/heartbeat:" + trigger.executionId,
+              payload: { scheduleKey: schedule.key },
+            });
+          }`,
+        });
+        break;
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/config-repo-template/worker.ts:65]: Fixed in 511e998b2. The raw seed commit still exits while creation is open, so it cannot add a second worker probe or hold the Project processor cursor ahead of `repos/created`. Instead, creation’s existing successful readiness probe now reads the trusted OS-stamped source commit and atomically appends the permanent feed, `project/created`, and the first `project/worker-updated`. The regression exercises `create-requested -> raw seed commit -> repos/created` and proves exactly one probe, one clean worker-update, and one creation terminal. The full repository typecheck, lint, format, and test gate passes.

File: apps/os/src/domains/projects/project-processor-implementation.ts [hunk 8]
```diff
@@ -361,51 +526,43 @@ export class ProjectProcessor extends StreamProcessor<
...
           }
-          return;
-        } finally {
-          // The returned Response can be a Cap'n Web RPC stub, and keeping
-          // that stub alive after the probe finishes is exactly the lifecycle
-          // pattern these stream tests are trying to avoid. Dispose on every
-          // attempt; local/miniflare Response objects without the hook are a
-          // no-op here.
-          disposeRpcResult(response);
+          return commitOid;
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/projects/project-processor-implementation.ts:587]: Confirmed. `project.worker.fetch` is capability dispatch, so its serialized userspace `Response` cannot supply the platform serve stamp. Fixed in 1c19a7b11: `ProjectDurableObject` now injects a direct `DynamicWorkerRunner.fetch` call into the Project processor, using the existing fetch lane that resolves/builds/loads the current default worker and stamps the trusted resolved commit after userspace returns. The harness now models that explicit dependency, and `worker-runner.test.ts` proves a userspace-spoofed serve header is replaced by the resolved repo commit. Focused processor/runner tests (45/45), OS typecheck/template typecheck, root lint, formatting, and diff checks pass.

File: apps/os/src/domains/projects/project-processor-implementation.ts [hunk 9]
```diff
@@ -109,25 +128,180 @@ export class ProjectProcessor extends StreamProcessor<
...
+              },
+            });
+            return;
+          }
+          const servedOutcomeIdempotencyKey = `project/worker-update:${servedCommitOid}`;
+          await append({
+            type: "events.iterate.com/project/worker-updated",
+            idempotencyKey: servedOutcomeIdempotencyKey,
+            payload: { commitOid: servedCommitOid },
+          });
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/projects/project-processor-implementation.ts:345]: Fixed in 776f7588d. Successful outcomes now keep the triggering config commit’s idempotency key while reporting the commit actually served. The regression test replays from a fresh processor checkpoint against a newly broken HEAD and proves it performs zero new probes and cannot append a contradictory failure.

File: apps/os/src/domains/projects/project-processor-contract.ts [hunk 10]
```diff
@@ -419,12 +499,109 @@ export type ProjectProcessorContract = typeof ProjectProcessorContract;
...
+      return parsed.success &&
+        parsed.data.createRequestedAtOffset === requestOffset &&
+        sameProjectCreationRequest(request, parsed.data.request)
+        ? { type: event.type, payload: parsed.data }
+        : null;
+    }
+    default:
+      return null;
+  }
+}
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/projects/project-processor-contract.ts:606]: Fixed in fe1ce335a. Creation terminals now use canonical `iterate-internal/project-creation-terminal` keys, which public appends reject before idempotency lookup. The Project DO commits the literal terminal batch through the existing server-only Stream stub, fenced to the stream lifetime that delivered the saga event; reduction validates the reserved key plus the exact open request instead of trusting forgeable `source.processor`. Added regressions for forged public processor-source claims, reserved-key rejection (including after the real event exists), and stream-lifetime mismatch. The exact repository gate (`pnpm install && pnpm typecheck && pnpm lint && pnpm format && pnpm test`) is green.

File: apps/os/src/domains/projects/project-processor-implementation.ts [hunk 11]
```diff
@@ -109,25 +135,204 @@ export class ProjectProcessor extends StreamProcessor<
...
+            if (
+              existingOutcome.type !== "events.iterate.com/project/worker-updated" &&
+              (existingOutcome.type !== "events.iterate.com/project/worker-update-failed" ||
+                existingOutcome.payload?.commitOid !== event.payload.commitOid)
+            ) {
+              throw new Error(
+                `idempotency key "${outcomeIdempotencyKey}" is not this config commit's worker update outcome`,
+              );
+            }
+            return;
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/projects/project-processor-implementation.ts:309]: Fixed in ffb7a9d15 (included in current head ccbfd6d6f). Both the seed and later worker-update outcomes now use the reserved internalStreamId("project-worker-update", commitOid) key and commit through the existing platform-only stream door. Public appends reject the iterate-internal namespace before idempotency lookup, so userspace cannot pre-claim the outcome; lost-checkpoint replay still finds the genuine committed result. Added the project-worker-update family to the public-append rejection regression. Verified with the full repository gate plus 71 focused ProjectProcessor/CoreProcessor tests.
